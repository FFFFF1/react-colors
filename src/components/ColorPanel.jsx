/* 选择色度和明度 */

// todo: 添加propTypes和defaultProps
// todo: 将div叠加修改为梯度背景色的叠加
// todo: 使用梯度和filter，添加其他二维选择功能
// todo: 客制化选框

import React, { Component, PureComponent, PropTypes } from 'react'
import reactCSS from 'reactcss'
import * as saturation from '../helpers/saturation'
import throttle from 'lodash/throttle'

export class ColorPanel extends (PureComponent || Component) {
  static propTypes = {
    color: PropTypes.string,
    colorModel: PropTypes.oneOf(['r','g','b','h','s','v']),
  }

  static defaultProps = {
    xAxis: 'h',
  }

  constructor(props) {
    super(props)

    // 50ms内不重复触发
    this.throttle = throttle((fn, data, e) => {
      fn(data, e)
    }, 50)
  }

  getBackground( model ) {
    let props = this.props
    let background = {
      r: `linear-gradient(to top right, rgba(${ props.rgb.r },0,0,1),transparent, rgba(${ props.rgb.r },255,255,1) ), 
          linear-gradient(to bottom right, rgb(${ props.rgb.r },255,0), rgb(${ props.rgb.r },0,255) )`,
      g: `linear-gradient(to top right, rgba(0,${ props.rgb.g },0,1),transparent, rgba(255,${ props.rgb.g },255,1) ), 
          linear-gradient(to bottom right, rgb(255,${ props.rgb.g },0), rgb(0,${ props.rgb.g },255) )`,
      b: `linear-gradient(to top right, rgba(0,0,${ props.rgb.b },1),transparent, rgba(255,255,${ props.rgb.b },1) ), 
          linear-gradient(to bottom right, rgb(0,255,${ props.rgb.b }), rgb(255,0,${ props.rgb.b }) )`,
      h: `linear-gradient(to top, #000, transparent),linear-gradient(to right, #FFF, rgba(255,255,255,0)),
          linear-gradient(to top, hsl(${ props.hsv.h }, 100%, 50%), hsl(${ props.hsv.h }, 100%, 50%))`,
      s: `linear-gradient(to top, #000, transparent), linear-gradient(rgba(255,255,255,${ 1-props.hsv.s }), rgba(255,255,255,${ 1-props.hsv.s })),
          linear-gradient(to right, #f00 0%, #ff0 17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)`,
      v: `linear-gradient(to top, rgba(0,0,0,${ 1-props.hsv.v }), rgba(0,0,0,${ 1-props.hsv.v })),linear-gradient(to top, #fff, transparent), 
          linear-gradient(to right, #f00 0%, #ff0 17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)`
    }

    return background[model]
  }



  componentWillUnmount() {
    this.unbindEventListeners()
  }

  handleChange = (e, skip) => {
    this.props.onChange(saturation.calculateChange(e, skip, this.props, this.refs.container), e)
  }

  handleMouseDown = (e) => {
    this.handleChange(e, true)
    window.addEventListener('mousemove', this.handleChange)
    window.addEventListener('mouseup', this.handleMouseUp)
  }

  handleMouseUp = () => {
    this.unbindEventListeners()
  }

  unbindEventListeners() {
    window.removeEventListener('mousemove', this.handleChange)
    window.removeEventListener('mouseup', this.handleMouseUp)
  }

  render() {
    let props = this.props
    const styles = reactCSS({
      'default': {
        root: {
          position: 'relative',
          display: 'inline-block',
          width: '256px',
          height: '256px',
          background: this.getBackground(props.colorModel),
        },
        white: {
          absolute: '0px 0px 0px 0px',
          background: 'linear-gradient(to right, #fff, rgba(255,255,255,0))',
        },
        black: {
          absolute: '0px 0px 0px 0px',
          background: 'linear-gradient(to top, #000, rgba(0,0,0,0))',
        },
        pointer: {
          position: 'absolute',
          top: `${ -(this.props.hsv.v * 100) + 100 }%`,
          left: `${ this.props.hsv.s * 100 }%`,
          cursor: 'default',
        },
        circle: {
          width: '4px',
          height: '4px',
          boxShadow: `0 0 0 1.5px #fff, inset 0 0 1px 1px rgba(0,0,0,.3),
            0 0 1px 2px rgba(0,0,0,.4)`,
          borderRadius: '50%',
          cursor: 'hand',
          transform: 'translate(-2px, -2px)',
        },
      },
      'custom': {
        ...this.props.style
      },
    }, { 'custom': !!this.props.style })

    return (
      <div
        style={ styles.root }
        ref="container"
        onMouseDown={ this.handleMouseDown }
        onTouchMove={ this.handleChange }
        onTouchStart={ this.handleChange }
      >
      </div>
    )
  }
}

export default ColorPanel