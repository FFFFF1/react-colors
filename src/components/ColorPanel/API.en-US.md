| Property       | Description           | Type             | Default       |
|------------|----------------|------------------|--------------|
| model      | picker model | rgb.r\|rgb.g\|rgb,b\|hsv.h\|hsv.s\|hsv.v | -  |
| onChange   | a callback function  | function    | -           |
| pointer    | user pointer element       | string\|ReactNode | <CirclePointer /> |
| color      | from ColorPicker | array    | -        |
| style      | container style       | object     | -           |
| pointerStyle | pointer style  | object  | -         |

