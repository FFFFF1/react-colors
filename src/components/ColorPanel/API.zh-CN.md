| 参数       | 说明           | 类型             | 默认值       |
|------------|----------------|------------------|--------------|
| model      | 拾色模式，未指定则由ColorPicker传入 | 之一: 'rgb.r' 'rgb.g' 'rgb.b' 'hsv.h' 'hsv.s' 'hsv.v' | 无  |
| onChange   | 变化回调，返回对应色彩模式的颜色值  | function    | 无           |
| pointer    | 指针，       | node | CirclePointer |
| color      | 颜色，由ColorPicker传入 | array    | 无        |
| style      | 容器样式       | object     | 无           |
| pointerStyle | 指针样式  | object  | 无         |