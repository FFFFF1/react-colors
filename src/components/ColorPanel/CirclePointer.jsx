import React, {Component, PropTypes} from 'react'

export default function CirclePointer(props) {
	let { style, ...p } = props
	const root = Object.assign({
    width: '4px',
    height: '4px',
    boxShadow: `0 0 0 1.5px #fff, inset 0 0 1px 1px rgba(0,0,0,.3),
      0 0 1px 2px rgba(0,0,0,.4)`,
    borderRadius: '50%',
    transform: 'translate(-2px, -2px)',
	}, style)

	return (
		<div style={ root } {...p}></div>
	)
}

CirclePointer.propTypes = {
	style: PropTypes.object
}
