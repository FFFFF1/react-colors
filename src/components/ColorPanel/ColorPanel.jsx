/* 选择色度和明度 */

// todo: 改变样式实现
// todo: 添加x，y轴配置

import React, { Component, PureComponent, PropTypes } from 'react'
import reactCSS from 'reactcss'
import calcEventPosition from '../../helpers/calcEventPosition'
import throttle from 'lodash/throttle'
import CirclePanel from './CirclePointer.jsx'

export class ColorPanel extends (PureComponent || Component) {
  static propTypes = {
    model: PropTypes.oneOf(['rgb.r','rgb.g','rgb.b','hsv.h','hsv.s','hsv.v']),
    pointer: PropTypes.node,
    rgb: PropTypes.object,
    hsv: PropTypes.object,
    onChange: PropTypes.func
  }

  static defaultProps = {
    pointer: <CirclePanel />
  }

  constructor(props) {
    super(props)

    // 50ms内不重复触发,todo 使用
    this.handleChange = throttle(this.handleChange, 50)
  }

  getBackground() {
    let props = this.props
    let background = {
      'rgb.r': `linear-gradient(to top right, rgba(${ Math.round(props.color[0]) },0,0,1),transparent, rgba(${ Math.round(props.color[0]) },255,255,1) ), 
          linear-gradient(to bottom right, rgb(${ Math.round(props.color[0]) },255,0), rgb(${ Math.round(props.color[0]) },0,255) )`,
      'rgb.g': `linear-gradient(to top right, rgba(0,${ Math.round(props.color[1]) },0,1),transparent, rgba(255,${ Math.round(props.color[1]) },255,1) ), 
          linear-gradient(to bottom right, rgb(255,${ Math.round(props.color[1]) },0), rgb(0,${ Math.round(props.color[1]) },255) )`,
      'rgb.b': `linear-gradient(to top right, rgba(0,0,${ Math.round(props.color[2]) },1),transparent, rgba(255,255,${ Math.round(props.color[2]) },1) ), 
          linear-gradient(to bottom right, rgb(0,255,${ Math.round(props.color[2]) }), rgb(255,0,${ Math.round(props.color[2]) }) )`,
      'hsv.h': `linear-gradient(to top, #000, transparent),linear-gradient(to right, #FFF, rgba(255,255,255,0)),
          linear-gradient(to top, hsl(${ props.color[0] }, 100%, 50%), hsl(${ props.color[0] }, 100%, 50%))`,
      'hsv.s': `linear-gradient(to top, #000, transparent), linear-gradient(rgba(255,255,255,${ 1-props.color[1] }), rgba(255,255,255,${ 1-props.color[1] })),
          linear-gradient(to right, #f00 0%, #ff0 17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)`,
      'hsv.v': `linear-gradient(to top, rgba(0,0,0,${ 1-props.color[2] }), rgba(0,0,0,${ 1-props.color[2] })),linear-gradient(to top, #fff, transparent), 
          linear-gradient(to right, #f00 0%, #ff0 17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)`
    }

    return background[props.model]
  }

  getColor(e) {
    let {color, model} = this.props
    let p = calcEventPosition(e, this.refs.container)
    let newColor = {
      'rgb.r': [color[0], 255 - p.topP * 255, p.leftP * 255],
      'rgb.g': [255 - p.topP * 255, color[1], p.leftP * 255],
      'rgb.b': [255 - p.topP * 255, p.leftP * 255, color[2]],
      'hsv.h': [color[0], p.leftP, 1 - p.topP],
      'hsv.s': [p.leftP * 360, color[1], 1 - p.topP],
      'hsv.v': [p.leftP * 360, 1 - p.topP, color[2]],
    }
    return newColor[model]
  }

  getPosition () {
    let {color, model} = this.props
    let position = {
      'rgb.r': {
        leftP: color[2] / 255,
        topP: 1 - color[1] / 255
      },
      'rgb.g': {
        leftP: color[2] / 255,
        topP: 1 - color[0] / 255
      },
      'rgb.b': {
        leftP: color[1] / 255,
        topP: 1 - color[0] / 255
      },
      'hsv.h': {
        leftP: color[1],
        topP: 1 - color[2]
      },
      'hsv.s': {
        leftP: color[0] / 360,
        topP: 1 - color[2]
      },
      'hsv.v': {
        leftP: color[0] / 360,
        topP: 1 - color[1]
      }
    }
    return position[model]
  }

  componentWillUnmount() {
    this.unbindEventListeners()
  }

  handleChange = (e) => {
    let c = this.getColor(e)
    this.props.onChange(c, e)
  }

  handleMouseDown = (e) => {
    this.handleChange(e, true)
    window.addEventListener('mousemove', this.handleChange)
    window.addEventListener('mouseup', this.handleMouseUp)
  }

  handleMouseUp = () => {
    this.unbindEventListeners()
  }

  unbindEventListeners() {
    console.log('remove')
    window.removeEventListener('mousemove', this.handleChange)
    window.removeEventListener('mouseup', this.handleMouseUp)
  }

  render() {
    let props = this.props
    let p = this.getPosition()
    const styles = reactCSS({
      'default': {
        root: {
          position: 'relative',
          display: 'inline-block',
          width: '256px',
          height: '256px',
          background: this.getBackground(),
          cursor: 'default',
        },
        pointer: {
          position: 'absolute',
          top: p.topP * 100 + '%',
          left: p.leftP * 100 + '%',
        },
        circle: {
          position: 'absolute',
          width: '4px',
          height: '4px',
          boxShadow: `0 0 0 1.5px #fff, inset 0 0 1px 1px rgba(0,0,0,.3),
            0 0 1px 2px rgba(0,0,0,.4)`,
          borderRadius: '50%',
          transform: 'translate(-2px, -2px)',
        },
      },
      'custom': {
        ...this.props.style
      },
    }, { 'custom': !!this.props.style })

    return (
      <div
        style={ styles.root }
        ref="container"
        onMouseDown={ this.handleMouseDown }
        onTouchMove={ this.handleChange }
        onTouchStart={ this.handleChange }
      >
        <div style={ styles.pointer }>{ 
          props.pointer
        }</div>
      </div>
    )
  }
}

export default ColorPanel