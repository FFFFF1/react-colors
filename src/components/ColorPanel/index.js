export {default} from './ColorPanel.jsx' 