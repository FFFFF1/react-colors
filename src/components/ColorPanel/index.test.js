import React from 'react'
import ColorPanel from './ColorPanel.jsx'
import { mount, render } from 'enzyme'
import { renderToJson } from 'enzyme-to-json'

describe('ColorPanel renders correctly', () => {
	it('render correctly', () => {
		const wrapper = render(<ColorPanel color={[255,0,0]} model='rgb.r'/>)
		expect(renderToJson(wrapper)).toMatchSnapshot()
	})

	it('show color correctly', () => {
		const wrapper = mount(<ColorPanel color={[255,127.5,127.5]} model='rgb.r'/>)
		expect(parseInt(wrapper.childAt(0).getNode().style.top)).toBe(50)
		expect(parseInt(wrapper.childAt(0).getNode().style.left)).toBe(50)
	})

	it('mousedown set color correctly', () => {
		const colorPanel = mount(<ColorPanel color={[255,0,0]} model='rgb.r' onChange={newColor => {
			newColor[1] = Math.round(newColor[1])
			newColor[2] = Math.round(newColor[2])
			expect(newColor).toEqual([255,155,100])
		}}/>).find(ColorPanel)

		colorPanel.getNode().container = {
			clientWidth: 256,
			clientHeight: 256,
			getBoundingClientRect: function () {
				return { left: 100, top: 100 }
			}
		}	
		colorPanel.simulate('mouseDown',{
			pageX: 200,
			pageY: 200
		})
	})
})


