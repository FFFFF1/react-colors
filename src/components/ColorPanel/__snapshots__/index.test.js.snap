// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPanel renders correctly render correctly 1`] = `
<div
  style="position:relative;height:100%;background:linear-gradient(to top right, rgba(255,0,0,1),transparent, rgba(255,255,255,1) ), 
          linear-gradient(to bottom right, rgb(255,255,0), rgb(255,0,255) );cursor:default;"
>
  <div
    style="position:absolute;top:100%;left:0%;"
  >
    <div
      style="width:4px;height:4px;box-shadow:0 0 0 1.5px #fff, inset 0 0 1px 1px rgba(0,0,0,.3),
          0 0 1px 2px rgba(0,0,0,.4);border-radius:50%;transform:translate(-2px, -2px);"
    />
  </div>
</div>
`;
